numberscope/backscope
# Clean up requirements

### Purpose

I want to add a new top-level requirement ([`structlog`](https://www.structlog.org/en/stable/why.html)) for the logging upgrade, so I cleaned up the current requirements first.

### Changes

- There were a bunch of packages in `requirements-freeze.txt` that don't seem to be used anymore. This PR removes them.
- There were some packages that seem to be top-level requirements, but weren't listed in `requirements.txt`. This PR adds them to `requirements.txt`.
- I wrote the script `tools/explain-requirements.sh` to help myself understand the dependency graph. It's documented in `requirements.md`.

### Review

Here's are the main review tasks I'd like people to try. Of course, you can also review in other ways!

#### Test the new requirements

1. Trash the current `.venv` directory.
2. Use `venv` to create a new virtual environment.
3. Call `sh tools/install-requirements.txt` to install the requirements in the updated `requirements-freeze.txt`.
4. Run automated and manual tests to make sure Backscope still works.

#### Test the requirements explanation tool

1. Look at the "Understanding dependencies" section of the updated `requirements.md`.
2. See if you can use that documentation to run the requirements explanation script and interpret its output.
